src/SnapKeyring.ts

```diff
@@ -397,68 +398,173 @@ export class SnapKeyring extends EventEmitter {
 
-    const response = await (async () => {
-      try {
-        return await this.#snapClient.withSnapId(snapId).submitRequest({
-          id: requestId,
-          scope: chainId,
-          account: account.id,
-          request: {
-            method,
-            ...(params !== undefined && { params }),
-          },
-        });
-      } catch (error) {
-        // If the snap failed to respond, delete the promise to prevent a leak.
-        this.#requests.delete(snapId, requestId);
-        throw error;
-      }
-    })();
+    const response = await this.#submitSnapRequest({
+      snapId,
+      requestId,
+      account,
+      method: method as EthMethod,
+      params,
+      chainId,
+    });
 
     // If the snap answers synchronously, the promise must be removed from the
     // map to prevent a leak.
     if (!response.pending) {
-      this.#requests.delete(snapId, requestId);
-      return response.result;
+      return this.#handleSyncResponse(response, requestId, snapId);
     }
 
     // If the snap answers asynchronously, we will inform the user with a redirect
     if (response.redirect?.message || response.redirect?.url) {
-      const { message = '', url = '' } = response.redirect;
-
-      // Check redirect url domain is in the snap allowed origins
-      if (url) {
-        const { origin } = new URL(url);
-        const snap = this.#snapClient.getController().get(snapId);
-        if (!snap) {
-          throw new Error(`Snap '${snapId}' not found.`);
-        }
-        const allowedOrigins = this.#getSnapAllowedOrigins(snap);
-        if (!allowedOrigins.includes(origin)) {
-          throw new Error(
-            `Redirect URL domain '${origin}' is not an allowed origin by snap '${snapId}'`,
-          );
-        }
-      }
-
-      await this.#callbacks.redirectUser(snapId, url, message);
+      await this.#handleAsyncResponse(response.redirect, snapId);
     }
 
     return promise.promise;
   }
 
+  /**
+   * Check if an account supports the given method.
+   *
+   * @param account - The account object to check for method support.
+   * @param method - The Ethereum method to validate.
+   * @returns `true` if the method is supported, `false` otherwise.
+   */
+  #hasMethod(account: KeyringAccount, method: EthMethod): boolean {
+    return account.methods.includes(method);
+  }
+
+  /**
+   * Creates a promise for a request and adds it to the map of requests.
+   *
+   * @param requestId - The unique identifier for the request.
+   * @param snapId - The Snap ID associated with the request.
+   * @returns A DeferredPromise instance.
+   */
+  #createRequestPromise<Response>(
+    requestId: string,
+    snapId: SnapId,
+  ): DeferredPromise<Response> {
+    const promise = new DeferredPromise<Response>();
+    this.#requests.set(requestId, { promise, snapId });
+    return promise;
+  }
+
+  /**
+   * Submits a request to a Snap and handles the response.
```

Comment:
Strictly speaking, we do not handles the response, as this is done by the caller with the respective `{Sync,Async}Response` helpers.

Unless you were talking about the error cases here?